cmd/assert.go

```diff
@@ -19,68 +19,79 @@ const (
 	SLOsByPrefix = "slosby"
 )
 
-// assert conditions
-var conds []string
-
 // how long to sleep in between retries of asserts
 var sleepTime, _ = time.ParseDuration("3s")
 
 // how long have we spent so far in assert attempts
 var timeSpent, _ = time.ParseDuration("0s")
 
-// timeout for assert conditions to be satisfied
-var timeout time.Duration
-
-// AssertCmd represents the assert command
-var AssertCmd = &cobra.Command{
-	Use:   "assert",
-	Short: "assert if experiment result satisfies the specified conditions",
-	Long: `
-	Assert if experiment result satisfies the specified conditions. 
-	If assert conditions are satisfied, exit with code 0. 
-	Else, return with code 1.`,
-	Example: `
-	# assert that the experiment completed without failures, 
-	# and SLOs were satisfied
-	iter8 assert -c completed -c nofailure -c slos
-
-	# another way to write the above assertion
-	iter8 assert -c completed,nofailure,slos
-
-	# if the experiment involves multiple app versions, 
-	# SLOs can be asserted for individual versions
-	# for example, the following command asserts that
-	# SLOs are satisfied by version numbered 0
-	iter8 assert -c completed,nofailures,slosby=0
-
-	# timeouts are useful for an experiment that may be long running
-	# and may run in the background
-	iter8 assert -c completed,nofailures,slosby=0 -t 5s
-	`,
-	RunE: func(cmd *cobra.Command, args []string) error {
-		// build experiment
-		exp := &Experiment{
-			Experiment: &base.Experiment{},
-		}
-		log.Logger.Trace("build started")
-		// replace FileExpIO with ClusterExpIO to build from cluster
-		fio := &FileExpIO{}
-		exp, err := Build(true, fio)
-		log.Logger.Trace("build finished")
-		if err != nil {
-			return err
-		}
+type AssertOptionsType struct {
+	// assert conditions
+	Conds []string
+	// timeout for assert conditions to be satisfied
+	Timeout time.Duration
+}
 
-		allGood, err := exp.Assert(conds, timeout)
-		if err != nil {
-			return err
-		}
-		if !allGood {
-			log.Logger.Error("assert conditions failed")
-			os.Exit(1)
-		}
-		return nil
-	},
+var AssertOptions = AssertOptionsType{}
+
+func NewAssertCmd() *cobra.Command {
```

Comment:
The `NewAssertCmd` function wrapper does nothing other than add needless indirection.

I think Cobra's default `AssertCmd` var works better than this wrapper which is adding complexity without value.

Reply:
Two instances of the `assert` command (and `run` and `report`) are needed. Both for `iter8 assert` and `iter8 k assert`.  It is not sufficient to reuse the same instance since the options and behaviors are different.  A `NewAssertCmd` method does this. For uniformity, I applied the same pattern everywhere.

Reply:
Let us please use the default Cobra way of handling  this using command variables.

Please rename `iter8 k assert` command as `KAssertCmd` or whatever you like, and use it along side `AssertCmd`. 

If you prefer that  I do this in a subsequent PR, I am take this activity on.

We want to use `cobra add`  command functions and be able to use the auto generated command scaffolding created by Cobra  with the least amount of edits. This function wrapper increases the  work needed to reuse Cobra's auto generated scaffolding. Is not really needed and an extra step in the way of creating new commands.

Reply:
I `git clone`d your repo and did the following.

```
cd iter8/k8s
cobra add test
```

At this point,  I got  a `test.go` file, nicely scaffolded, under `iter8/k8s/cmd`.

It has `var testCmd` initialized to a command inside it.

***

What the above means is that you can simply do `cobra add assert`, cobra add run` etc.

You will get a runCmd, which you can add as a child to `basecli.RunCmd`, `basecli.AssertCmd` and so on.

***

There is absolutely no conflict. In one case, you have `assertCmd` which is a package scoped variable inside the `k8s` package. In another case, you have an `AssertCmd`, which is an exported variable inside the `basecli` package.

***

All this without having to do the extra work of recreating what cobra already creates for you using its builtin `add` functionality.

Reply:
Let me reiterate that I use this pattern to enable reuse. This eliminates the need to re-implement or duplicate code. I do not redefine all the flags that `cmd/assert.go` defines. Nor do I redefine the name, example or description. Instead, they are defined just once and reused. This is accomplished by using two calls to `NewAssertCmd()` -- one in `cmd/assert.go` and one in `k8s/cmd/assert.go`.

This pattern may not exactly adhere to the "simplest" cobra pattern. That does not mean that it isn't a valid and useful pattern. The differences are minor and they provide significant reuse in a few cases.

A requirement to support `cobra add` as a way to **stub** code is a new requirement. It is not at all clear that the benefits are worth any effort. In particular, any commands that support k8s generic flags need to do a bunch of extra work and will require a rewrite of the stub. That said, naming is irrelevant. Two small changes enabled this feature.